bin/test-run

```
@@ -81,29 +94,34 @@ k8s_context=${3:-""}
 
 check_linkerd_binary
 check_if_k8s_reachable
+remove_l5d_if_exists
 
 printf "==================RUNNING ALL TESTS==================\\n"
-
 printf "Testing Linkerd version [%s] namespace [%s] k8s-context [%s]\\n" "$linkerd_version" "$linkerd_namespace" "$k8s_context"
 
-exit_code=0
-
+# run upgrade test:
+# 1. install latest stable
+# 2. upgrade to HEAD
+# 3. if failed, exit script to avoid leaving behind stale resources which will
+# fail subsequent tests. `cleanup` is not called if this test failed so that
+# there is a chance to debug the problem
 run_upgrade_test "$linkerd_namespace"-upgrade || exit_code=$?
-# if run_upgrade_test succeeded, delete the installation to prepare for subsequent
-# full integration tests.
-# TODO: Consider running the upgrade tests and normal integration tests on
-# separate clusters (in parallel? via kind?).
-if [ $exit_code -eq 0 ]; then
-    $bindir/test-cleanup $k8s_context
+if [ $exit_code -ne 0 ]; then
+  printf "\\n=== FAIL: Can't upgrade to version $linkerd_version\\n"
+  exit $exit_code
 fi
+cleanup
 
-run_test "$test_directory/install_test.go" --linkerd-namespace=$linkerd_namespace || exit_code=$?
+# run the rest of the test suite. `cleanup` is not called if this test failed
+# so that there is a chance to debug the problem
+run_test "$test_directory/install_test.go" -failfast --linkerd-namespace=$linkerd_namespace || exit_code=$?
```

Fail fast for `install_test.go`. No point of continuing.